### PR TITLE
Big update

### PR DESCRIPTION
**Pull request by gabrieltseng**
This PR introduces a number of significant updates:

- [x] **Model updates**: Presto previously considered SRTM as dynamic in time (`num_timesteps` SRTM tokens were passed through the model), even though it was only collected at a single timestep. A single SRTM token is now passed through the model. In addition, a bug in the decoder which shuffled some bands during reconstruction has been fixed.
- [x] **Masking updates**: The masking functions were updated to work with both dynamic-in-time and static-in-time inputs. In addition, they were updated to accept arbitrary masking ratios (previously, the masking ratio had to yield a number of masked tokens divisible by the number of timesteps and the number of band groups).

 The `default_model.pt` weights have now been updated with a model trained after these changes.

- [x] **Evaluation updates (images)**: For evaluation tasks which make a single prediction for an image, we now pass a mean and standard deviation of the pixel-outputs for that image to a downstream classifier (instead of passing each individual pixel to the classifier and taking the mode of the classifier's outputs).
- [x] **Evaluation updates (seeds)**: All downstream classifiers with randomness (i.e. everything except the K-nearest-neighbours classifier and the linear regression) are run for 3 seeds to measure Presto's sensitivity to seeds.  
- [x] **Evaluation updates (EuroSat)**:  Our EuroSat evaluation now better mirrors the approach taken by previous works; we downsample images to different resolutions, and process all the pixels in the image. In addition, we use the splits described #8). 

A manuscript with updated results will be posted on arxiv.

- [x] **Many more tests**: We add many more tests to ensure our masking functions are doing what we expect, and to ensure tokens are being handled in the right way by the Presto encoder and decoder.
- [x] **Other infrastructure updates**:  Lots of changes here: (i) we add logging to the training and evaluation code for clarity, (ii) speed up evaluation tasks by using multiple workers for the eval dataloaders, (iii) replace the processing code for TreeSat and EuroSat with a dataloader which reads the raw files (so there is no need to process the data before running th evaluation code), (iv) run CropHarvest validation less often to speed up pre-training, (iv) make the `data_dir` configurable in `train.py`, (vi) dump the evaluation results in a json file locally in addition to (optionally) storing them on wandb.

The repository before this PR is tagged `v0.1`.

### CODE REVIEW

File: presto/eval/eval.py [hunk 1]
```diff
@@ -1,25 +1,46 @@
...
+
+
+class EvalTask(ABC):
     name: str
     num_outputs: int
     regression: bool
     multilabel: bool
 
+    def __init__(self, seed: int = DEFAULT_SEED):
+        self.seed = seed
```
Comment (rubencart) [inline presto/eval/eval.py:41]: I don't really like that this class has a `seed: int` argument and its subclass EvalTaskWithAggregatedOutput has a `seeds: List[int]` argument, so you basically have to use `isinstance(EvalTaskWithAggregatedOutput,...)` to know what field to use. I think it would be nicer to have this class use a field `seeds: List[int]` too and include a check this its length equals 1 (or something along these lines)

File: eval.py [hunk 2]
```diff
@@ -4,77 +4,195 @@
...
+        ]
+    )
+
+if wandb_enabled:
+    eval_config = {
+        "model": model.__class__,
+        "encoder": model.encoder.__class__,
+        "decoder": model.decoder.__class__,
+        "device": device,
+        "model_parameters": "random" if fully_supervised else path_to_state_dict,
```
Comment (rubencart) [inline eval.py:167]: I think I should have added `"logging_dir": logging_dir,` here but I forgot, same in train.py. Because now `output_dir` gets logged but it's basically always something like `/network/scratch/<u>/<user>/presto/` and then you don't know what the actual subdir for this run was

File: train.py [hunk 3]
```diff
@@ -194,7 +244,7 @@ def load_dataset(url, shuffle_on_load):
 }
```
Comment (rubencart) [inline train.py:244]: Add `"logging_dir": logging_dir,`

File: eval.py [hunk 4]
```diff
@@ -4,77 +4,195 @@
...
+)
+argparser.add_argument(
+    "--output_dir",
+    type=str,
+    default="",
+    help="Output is stored in <data_dir>/output. "
+    "Leave empty to use the directory you are running this file from.",
+)
 argparser.add_argument("--fully_supervised", dest="fully_supervised", action="store_true")
+argparser.add_argument("--wandb", dest="wandb", action="store_true")
```
Comment (rubencart) [inline eval.py:56]: Could you add an argument `argparser.add_argument("--nb_seeds", type=int, default=1)` so it's actually also still possible to run the eval script with just 1 seed? 
Or do we not want to support that?
Then below: `seeds = list(range(0, DEFAULT_SEED*nb_seeds, DEFAULT_SEED) if nb_seeds > 1 else [DEFAULT_SEED]`

File: eval.py [hunk 5]
```diff
@@ -4,77 +4,195 @@
...
+    TreeSatEval("S2", input_patch_size=6),
     CropHarvestEval("Kenya"),
     CropHarvestEval("Togo"),
     CropHarvestEval("Brazil"),
+    CropHarvestEval("Kenya", seed=0),
+    CropHarvestEval("Togo", seed=0),
+    CropHarvestEval("Brazil", seed=0),
+    CropHarvestEval("Kenya", seed=84),
+    CropHarvestEval("Togo", seed=84),
+    CropHarvestEval("Brazil", seed=84),
```
Comment (rubencart) [inline eval.py:144]: Then here 
```python
eval_task_list: List[EvalTask] = [
    *[CropHarvestEval("Kenya", seed=seed) for seed in seeds],
    *[CropHarvestEval("Kenya", ignore_dynamic_world=True, seed=seed) for seed in seeds],
    *[CropHarvestEval("Brazil", seed=seed) for seed in seeds],
    *[CropHarvestEval("Brazil", ignore_dynamic_world=True, seed=seed) for seed in seeds],
    *[CropHarvestEval("Togo", seed=seed) for seed in seeds],
    *[CropHarvestEval("Togo", ignore_dynamic_world=True, seed=seed) for seed in seeds],
    *[FuelMoistureEval(seed=seed) for seed in seeds],
    *[AlgaeBloomsEval(seed=seed) for seed in seeds],
    *[TreeSatEval("S1", input_patch_size=ps, seeds=seeds) for ps in [1, 2, 3, 6]],
    *[TreeSatEval("S2", input_patch_size=ps, seeds=seeds) for ps in [1, 2, 3, 6]],
    ... # EuroSat
]
```

File: eval.py (hunk 2 above)
Comment (gabrieltseng) [reply eval.py:159]: added!

File: presto/eval/eval.py (hunk 1 above)
Comment (gabrieltseng) [reply presto/eval/eval.py]: updated!

File: train.py (hunk 3 above)
Comment (gabrieltseng) [reply train.py:253]: updated

File: eval.py (hunk 4 above)
Comment (gabrieltseng) [reply eval.py:63]: added an `eval_seeds` arguments

File: eval.py (hunk 5 above)
Comment (gabrieltseng) [reply eval.py]: nice this is great. updated.

**Review by rubencart — APPROVED** [review]
Great, looks good to me!
